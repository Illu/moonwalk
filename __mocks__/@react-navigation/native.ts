import { Theme, DefaultTheme as defaultTheme } from "@react-navigation/native";

export const LightTheme = {
  ...defaultTheme,
  colors: {
    ...defaultTheme.colors,
    secondary: "#000",
    background: "#f1f2f7",
    text: "#000",
    placeholderText: "#7d7f86",
    inputBackground: "#dee1e7",
    secondaryText: "#666",
    accent: "#0a84ff",
    accentBackground: "rgba(10, 132, 255, 0.1)",
    uiAccent: "#c3c4c6",
  },
};

export const DefaultTheme = LightTheme;

export const DarkTheme: Theme = LightTheme;

export const useTheme = jest.fn().mockReturnValue({
  LightTheme,
});

export const useFocusEffect = jest.fn();
export const useNavigation = jest.fn().mockReturnValue({
  navigate: jest.fn(),
  replace: jest.fn(),
  reset: jest.fn(),
  setOptions: jest.fn(),
});
export const useNavigationState = jest.fn();
export const useRoute = jest.fn().mockReturnValue({
  params: {},
});
