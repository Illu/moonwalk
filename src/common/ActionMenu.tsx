import { useTheme } from "@react-navigation/native";
import React from "react";
import { View } from "react-native";
import styled from "styled-components/native";

import Icon from "./Icon";

// --ActionMenu--
// Takes an items object to display an organised menu.

const Wrapper = styled.View`
  background: ${({ theme }) => theme.colors.background};
`;

const Row = styled.TouchableOpacity`
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding: 20px 16px 20px 0;
  border-color: ${({ theme }) => theme.colors.uiAccent};
`;

const PreviewText = styled.Text`
  font-size: 16px;
  margin-right: 5px;
  color: ${({ theme }) => theme.colors.uiAccent};
`;

const Title = styled.Text`
  font-size: 16px;
  color: ${({ theme }) => theme.colors.text};
`;

const CategoryWrapper = styled.View`
  margin: 20px;
  border-radius: 10px;
  padding: 0 0 0 16px;
  background: ${({ theme }) => theme.colors.secondary};
`;

const ThumbWrapper = styled.View`
  width: 30px;
  height: 30px;
  margin-right: 10px;
  border-radius: 5px;
  align-items: center;
  justify-content: center;
`;

const ThumbImage = styled.Image`
  border-radius: 18px;
  height: 70px;
  width: 70px;
  margin-right: 20px;
`;

type Item = {
  title: string;
  icon: string;
  action?: () => any;
  preview?: string;
  thumbIcon?: string;
  thumbColor?: string;
  disabled?: boolean;
  thumbImage?: any;
};

interface Props {
  items: Array<Item[]>;
}

const ActionMenu = ({ items }: Props) => {
  const { colors } = useTheme();

  return (
    <Wrapper>
      {items.map((subItems, index) => (
        <CategoryWrapper key={index}>
          {subItems.map((item, i) => (
            <Row
              key={`${item.title}${i}`}
              style={{
                borderTopWidth: i !== 0 ? 0.5 : 0,
              }}
              onPress={item.action}
              disabled={item.disabled}
            >
              <View style={{ flexDirection: "row", alignItems: "center" }}>
                {item.thumbIcon && (
                  <ThumbWrapper style={{ backgroundColor: item.thumbColor }}>
                    <Icon name={item.thumbIcon} size={18} />
                  </ThumbWrapper>
                )}
                {item.thumbImage && (
                  <ThumbImage
                    source={
                      typeof item.thumbImage === "string"
                        ? { uri: item.thumbImage }
                        : item.thumbImage
                    }
                  />
                )}
                <Title>{item.title}</Title>
              </View>
              <View style={{ flexDirection: "row", alignItems: "center" }}>
                {item.preview && <PreviewText>{item.preview}</PreviewText>}
                <Icon name={item.icon} color={colors.uiAccent} size={25} />
              </View>
            </Row>
          ))}
        </CategoryWrapper>
      ))}
    </Wrapper>
  );
};

export default ActionMenu;
