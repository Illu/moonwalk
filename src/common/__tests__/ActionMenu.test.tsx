import * as navigation from "@react-navigation/native";
import React from "react";

import TestRenderer from "../../helpers/testRenderer";
import { lightTheme } from "../../theme";
import ActionMenu from "../ActionMenu";

jest.mock("../Icon", () => "Icon");
jest.spyOn(navigation, "useTheme");
navigation.useTheme.mockImplementation(() => lightTheme);

describe("ActionMenu", () => {
  it("renders correctly", () => {
    const props = {
      items: [
        [
          {
            title: "$_TEST_TITLE_$",
            icon: "$_TEST_ICON_$",
          },
          {
            title: "$_TEST_TITLE_$",
            icon: "$_TEST_ICON_$",
            disabled: true,
          },
        ],
        [
          {
            title: "$_TEST_TITLE_$",
            icon: "$_TEST_ICON_$",
            thumbColor: "tomato",
          },
          {
            title: "$_TEST_TITLE_$",
            icon: "$_TEST_ICON_$",
            preview: "$_PREVIEW_TEXT_$",
          },
          {
            title: "$_TEST_TITLE_$",
            icon: "$_TEST_ICON_$",
            action: jest.fn(),
          },
        ],
        [
          {
            title: "$_TEST_TITLE_$",
            icon: "$_TEST_ICON_$",
            thumbIcon: "$_TEST_THUMB_ICON_$",
          },
          {
            title: "$_TEST_TITLE_$",
            icon: "$_TEST_ICON_$",
            thumbImage: "$_TEST_THUMB_IMAGE_$",
          },
          {
            title: "$_TEST_TITLE_$",
            icon: "$_TEST_ICON_$",
            thumbImage: 1,
          },
        ],
      ],
    };
    const tree = TestRenderer(<ActionMenu {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
