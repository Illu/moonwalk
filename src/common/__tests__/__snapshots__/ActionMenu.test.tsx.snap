// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionMenu renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#f1f2f7",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 10,
          "marginBottom": 20,
          "marginLeft": 20,
          "marginRight": 20,
          "marginTop": 20,
          "paddingBottom": 0,
          "paddingLeft": 16,
          "paddingRight": 0,
          "paddingTop": 0,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#c3c4c6",
          "borderTopWidth": 0,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "opacity": 1,
          "paddingBottom": 20,
          "paddingLeft": 0,
          "paddingRight": 16,
          "paddingTop": 20,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 16,
              },
            ]
          }
        >
          $_TEST_TITLE_$
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Icon
          color="#c3c4c6"
          name="$_TEST_ICON_$"
          size={25}
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#c3c4c6",
          "borderTopWidth": 0.5,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "opacity": 1,
          "paddingBottom": 20,
          "paddingLeft": 0,
          "paddingRight": 16,
          "paddingTop": 20,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 16,
              },
            ]
          }
        >
          $_TEST_TITLE_$
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Icon
          color="#c3c4c6"
          name="$_TEST_ICON_$"
          size={25}
        />
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 10,
          "marginBottom": 20,
          "marginLeft": 20,
          "marginRight": 20,
          "marginTop": 20,
          "paddingBottom": 0,
          "paddingLeft": 16,
          "paddingRight": 0,
          "paddingTop": 0,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#c3c4c6",
          "borderTopWidth": 0,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "opacity": 1,
          "paddingBottom": 20,
          "paddingLeft": 0,
          "paddingRight": 16,
          "paddingTop": 20,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 16,
              },
            ]
          }
        >
          $_TEST_TITLE_$
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Icon
          color="#c3c4c6"
          name="$_TEST_ICON_$"
          size={25}
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#c3c4c6",
          "borderTopWidth": 0.5,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "opacity": 1,
          "paddingBottom": 20,
          "paddingLeft": 0,
          "paddingRight": 16,
          "paddingTop": 20,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 16,
              },
            ]
          }
        >
          $_TEST_TITLE_$
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#c3c4c6",
                "fontSize": 16,
                "marginRight": 5,
              },
            ]
          }
        >
          $_PREVIEW_TEXT_$
        </Text>
        <Icon
          color="#c3c4c6"
          name="$_TEST_ICON_$"
          size={25}
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#c3c4c6",
          "borderTopWidth": 0.5,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "opacity": 1,
          "paddingBottom": 20,
          "paddingLeft": 0,
          "paddingRight": 16,
          "paddingTop": 20,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 16,
              },
            ]
          }
        >
          $_TEST_TITLE_$
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Icon
          color="#c3c4c6"
          name="$_TEST_ICON_$"
          size={25}
        />
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 10,
          "marginBottom": 20,
          "marginLeft": 20,
          "marginRight": 20,
          "marginTop": 20,
          "paddingBottom": 0,
          "paddingLeft": 16,
          "paddingRight": 0,
          "paddingTop": 0,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#c3c4c6",
          "borderTopWidth": 0,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "opacity": 1,
          "paddingBottom": 20,
          "paddingLeft": 0,
          "paddingRight": 16,
          "paddingTop": 20,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 5,
                "height": 30,
                "justifyContent": "center",
                "marginRight": 10,
                "width": 30,
              },
              Object {
                "backgroundColor": undefined,
              },
            ]
          }
        >
          <Icon
            name="$_TEST_THUMB_ICON_$"
            size={18}
          />
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 16,
              },
            ]
          }
        >
          $_TEST_TITLE_$
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Icon
          color="#c3c4c6"
          name="$_TEST_ICON_$"
          size={25}
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#c3c4c6",
          "borderTopWidth": 0.5,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "opacity": 1,
          "paddingBottom": 20,
          "paddingLeft": 0,
          "paddingRight": 16,
          "paddingTop": 20,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "$_TEST_THUMB_IMAGE_$",
            }
          }
          style={
            Array [
              Object {
                "borderRadius": 18,
                "height": 70,
                "marginRight": 20,
                "width": 70,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 16,
              },
            ]
          }
        >
          $_TEST_TITLE_$
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Icon
          color="#c3c4c6"
          name="$_TEST_ICON_$"
          size={25}
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#c3c4c6",
          "borderTopWidth": 0.5,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "opacity": 1,
          "paddingBottom": 20,
          "paddingLeft": 0,
          "paddingRight": 16,
          "paddingTop": 20,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Image
          source={1}
          style={
            Array [
              Object {
                "borderRadius": 18,
                "height": 70,
                "marginRight": 20,
                "width": 70,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "#000",
                "fontSize": 16,
              },
            ]
          }
        >
          $_TEST_TITLE_$
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Icon
          color="#c3c4c6"
          name="$_TEST_ICON_$"
          size={25}
        />
      </View>
    </View>
  </View>
</View>
`;
